libraries/AP_Scheduler/AP_Scheduler.cpp

```diff
@@ -89,96 +90,87 @@ void AP_Scheduler::tick(void)
-                             (unsigned)_task_time_allowed);
-                }
-            }
 
-            if (_task_time_allowed <= time_available) {
-                // run it
-                _task_time_started = now;
-                current_task = i;
-                if (_debug > 1 && _perf_counters && _perf_counters[i]) {
-                    hal.util->perf_begin(_perf_counters[i]);
-                }
-                _tasks[i].function();
-                if (_debug > 1 && _perf_counters && _perf_counters[i]) {
-                    hal.util->perf_end(_perf_counters[i]);
-                }
-                current_task = -1;
-
-                // record the tick counter when we ran. This drives
-                // when we next run the event
-                _last_run[i] = _tick_counter;
-
-                // work out how long the event actually took
-                now = AP_HAL::micros();
-                uint32_t time_taken = now - _task_time_started;
-
-                if (time_taken > _task_time_allowed) {
-                    // the event overran!
-                    if (_debug > 4) {
-                        ::printf("Scheduler overrun task[%u-%s] (%u/%u)\n",
-                                 (unsigned)i,
-                                 _tasks[i].name,
-                                 (unsigned)time_taken,
-                                 (unsigned)_task_time_allowed);
-                    }
-                }
-                if (time_taken >= time_available) {
-                    goto update_spare_ticks;
-                }
-                time_available -= time_taken;
+        // run it
+        if (_task_time_allowed <= time_available) {
+            _task_time_started = task_time_buf;
+            current_task = i;
+            const bool bPerfCntrs = _debug > 1 && _perf_counters && _perf_counters[i];
+            if (bPerfCntrs) {
+                hal.util->perf_begin(_perf_counters[i]);
+            }
+            _tasks[i].function();
+            if (bPerfCntrs) {
+                hal.util->perf_end(_perf_counters[i]);
+            }
+            current_task = -1;
+
+            // record the tick counter when we ran. This drives
+            // when we next run the event
+            _last_run[i] = _tick_counter;
+
+            // work out how long the event actually took
+            task_time_buf = AP_HAL::micros();
+            const uint32_t time_taken = task_time_buf - _task_time_started;
+
+            // the event overran!
+            if (_debug > 4 && time_taken > _task_time_allowed) {
+                ::printf("Scheduler overrun task[%u-%s] (%u/%u)\n",
+                            (unsigned)i,
+                            _tasks[i].name,
+                            (unsigned)time_taken,
+                            (unsigned)_task_time_allowed);
             }
+            
+            if (time_taken >= time_available) {
+                time_available = 0;
```

Comment:
The goto is AFIAK more eficient, the the assignment and the break.

Reply:
Not really: https://godbolt.org/g/9veakW

Reply:
@dgrat you miss the count = 0; on the goto version, and the ouput is the same ! 
But I tottally agree that we should remove the goto, most people don't understand it, and it isn't needed here.

Reply:
Yes, there is one "mov 0" more and a jump label less.

Reply:
These are based on gcc 7.2 but AFAIK we are using gcc 5.4

Reply:
@amilcarlucas gcc 4.9 and ARM, but is the same but less optimised!